server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * IndexReshardingState is an abstract class holding the persistent state of a generic resharding operation. It contains
+ * concrete subclasses for the operations that are currently defined (which is only split for now).
+ */
+public abstract sealed class IndexReshardingState implements Writeable, ToXContentFragment {
+    // This class exists only so that tests can check that IndexReshardingMetadata can support more than one kind of operation.
+    // When we have another real operation such as Shrink this can be removed.
+    public static final class Noop extends IndexReshardingState {
+        private static final ObjectParser<Noop, Void> NOOP_PARSER = new ObjectParser<>("noop", Noop::new);
+
+        Noop() {}
+
+        Noop(StreamInput in) throws IOException {
+            this();
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            return builder;
+        }
+
+        static IndexReshardingState fromXContent(XContentParser parser) throws IOException {
+            return NOOP_PARSER.parse(parser, null);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {}
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            return other != null && getClass() == other.getClass();
+        }
+
+        @Override
+        public int hashCode() {
+            return 0;
+        }
+    }
+
+    public static final class Split extends IndexReshardingState {
+        public enum SourceShardState implements Writeable {
+            SOURCE,
+            DONE;
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
```

Comment:
I think we also want a constructor with `StreamInput` then? Can be follow-up for sure if it is not needed here.

Reply:
When I add the explicit number then I'll also add a static `readFrom` as in the example you linked. In this iteration it wasn't valuable since the parent was just calling `readEnum` inside `readArray`.

Reply:
Done in 0e69a76